ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendSpec.scala

```diff
@@ -5,67 +5,86 @@ package com.daml.platform.store.backend
 
 import java.sql.Connection
 import java.util.concurrent.atomic.AtomicInteger
-
-import com.codahale.metrics.MetricRegistry
-import com.daml.ledger.api.testing.utils.AkkaBeforeAndAfterAll
-import com.daml.ledger.resources.{Resource, ResourceContext}
+import com.daml.ledger.resources.ResourceContext
 import com.daml.logging.{ContextualizedLogger, LoggingContext}
-import com.daml.metrics.Metrics
-import com.daml.platform.configuration.ServerRole
 import com.daml.platform.store.FlywayMigrations
-import com.daml.platform.store.appendonlydao.DbDispatcher
-import org.scalatest.{AsyncTestSuite, BeforeAndAfterEach}
+import org.scalatest.{TestSuite, BeforeAndAfterAll, BeforeAndAfterEach}
 
-import scala.concurrent.duration.{DurationInt, FiniteDuration}
-import scala.concurrent.{Await, Future}
+import java.util.concurrent.Executors
+import javax.sql.DataSource
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.Try
 
 private[backend] trait StorageBackendSpec
-    extends AkkaBeforeAndAfterAll
+    extends StorageBackendProvider
     with BeforeAndAfterEach
-    with StorageBackendProvider { this: AsyncTestSuite =>
+    with BeforeAndAfterAll { this: TestSuite =>
 
   protected val logger: ContextualizedLogger = ContextualizedLogger.get(getClass)
   implicit protected val loggingContext: LoggingContext = LoggingContext.ForTesting
 
-  private val connectionPoolSize: Int = 16
-  private val metrics = new Metrics(new MetricRegistry)
+  // Data source (initialized once)
+  private var dataSource: DataSource = _
+
+  // Default connection for database operations (initialized for each test, since connections are stateful)
+  private var defaultConnection: Connection = _
 
-  // Initialized in beforeAll()
-  private var dbDispatcherResource: Resource[DbDispatcher] = _
-  private var dbDispatcher: DbDispatcher = _
+  // Execution context with a fixed number of threads, used for running parallel queries (where each query is
+  // executed in its own thread).
+  private val connectionPoolSize = 16
+  private val connectionPoolExecutionContext = ExecutionContext.fromExecutor(
+    Executors.newFixedThreadPool(
+      connectionPoolSize
+    )
+  )
+
+  /** Runs the given database operations in parallel.
+    * Each operation will run in a separate thread and will use a separate database connection.
+    */
+  protected def executeParallelSql[T](fs: Vector[Connection => T]): Vector[T] = {
```

Comment:
Almost all tests are sequential, however the deduplication tests do execute deduplication queries in parallel, to assert that they work correctly when multiple identical commands are submitted at the same time. A combination of `Future.sequence` and a dedicated fixed thread pool is used to attempt to run these queries as concurrently as possible.